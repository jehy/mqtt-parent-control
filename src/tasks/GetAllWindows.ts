import { parse } from 'csv-parse/sync';

import GetByShell from './GetByShell.ts';
import isWinPlatform from '../lib/isWin.ts';

import type { TaskOptions, TaskType } from './Task.ts';

export type GetAllWindowsConfig = {
  topic: string,
  user: string,
};

const winExtraTitles = ['Task', 'Host', 'Windows Push Notifications',
  'Platform', 'Начальный экран', 'OleMainThread', 'N/A', 'Н/Д',
  'CrossDeviceResumeWindow', 'OLEChannelWnd', 'Quick Settings', 'Search', 'Поиск', 'Пуск'];

export default class GetAllWindows extends GetByShell {
  public config: GetAllWindowsConfig;

  public name: TaskType = 'GetAllWindows';

  constructor(config: GetAllWindowsConfig, options: TaskOptions) {
    super(config, options);
    this.config = config;
    if (!this.config.user) {
      this.enabled = false;
      this.logs.push('user not found');
    }
    if (isWinPlatform) {
      this.command = 'tasklist  /v /nh /fo csv';
    } else {
      this.command = `su - ${config.user} -c  'DISPLAY=":0" wmctrl -l'`;
    }
  }

  public async end(): Promise<void> {
    const result = this.parseReply(this.shellResult.stdout.trim(), isWinPlatform);
    if (this.client) {
      await this.client.publish(this.config.topic, result);
    }
  }

  public parseReply(data: string, isWinPlatform: boolean): string {
    if (isWinPlatform) {
      if (data.split(',').length < 5) {
        return '';
      }
      const records = parse(data, {
        // "Имя образа","PID","Имя сессии","№ сеанса","Память","Состояние","Пользователь","Время ЦП","Заголовок окна"
        // eslint-disable-next-line no-irregular-whitespace
        // "powershell.exe","12832","RDP-Tcp#5","1","77 760 КБ","Running","BIG-PC\jehy","0:00:01","Администратор: Windows PowerShell"
        columns: ['name', 'pid', 'sessionName', 'sessionNum', 'memory', 'state', 'user', 'cpuTime', 'title'],
        skip_empty_lines: true,
      }) as Array<{ title: string, user: string }>;
      return records
        .filter((rec) => rec
          && rec.title && !winExtraTitles.find((el) => rec.title.includes(el))
          && rec.user.includes(this.config.user))
        .map((rec) => rec.title).sort().join(' | ');
    }
    return data.split('\n')
      .map((el) => {
        const all = el.split(' ');
        return all.slice(3).join(' ');
      })
      .sort()
      .join(' ')
      .substring(0, 255);
  }
}
