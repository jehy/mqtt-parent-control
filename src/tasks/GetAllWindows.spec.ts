import GetAllWindows from './GetAllWindows';

import type { TaskOptions } from './Task';

describe('GetAllWindows', () => {
  it('should be able to parse on WIN', async () => {
    let result: string = 'initValue';
    const fakeClient = {
      publish: (_topic: string, taskResult: string) => {
        result = taskResult;
      },
    };
    const options: TaskOptions = { client: fakeClient } as TaskOptions;
    const task = new GetAllWindows({ user: 'jehy', topic: 'windows' }, options);
    const reply = `
"System Idle Process","0","Services","0","8 КБ","Unknown","NT AUTHORITY\СИСТЕМА","2:39:14","Н/Д"
"System","4","Services","0","180 КБ","Unknown","Н/Д","0:00:10","Н/Д"
"Secure System","188","Services","0","99 692 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"Registry","232","Services","0","63 208 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"smss.exe","596","Services","0","1 700 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"csrss.exe","752","Services","0","5 900 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"wininit.exe","828","Services","0","8 716 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"csrss.exe","836","","1","6 616 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"winlogon.exe","904","","1","15 696 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"services.exe","976","Services","0","11 188 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"LsaIso.exe","996","Services","0","4 552 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"lsass.exe","1004","Services","0","22 180 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:01","Н/Д"
"svchost.exe","572","Services","0","40 236 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:01","Н/Д"
"fontdrvhost.exe","1052","","1","5 088 КБ","Unknown","Font Driver Host\UMFD-1","0:00:00","Н/Д"
"fontdrvhost.exe","1060","Services","0","4 664 КБ","Unknown","Font Driver Host\UMFD-0","0:00:00","Н/Д"
"svchost.exe","1132","Services","0","15 856 КБ","Unknown","NT AUTHORITY\NETWORK SERVICE","0:00:01","Н/Д"
"dwm.exe","1252","","1","76 320 КБ","Unknown","Window Manager\DWM-1","0:00:08","Н/Д"
"svchost.exe","1348","Services","0","7 016 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"svchost.exe","1356","Services","0","16 024 КБ","Unknown","NT AUTHORITY\NETWORK SERVICE","0:00:00","Н/Д"
"svchost.exe","1504","Services","0","28 448 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:01","Н/Д"
"svchost.exe","1532","Services","0","28 356 КБ","Unknown","NT AUTHORITY\LOCAL SERVICE","0:00:00","Н/Д"
"svchost.exe","1540","Services","0","30 484 КБ","Unknown","NT AUTHORITY\LOCAL SERVICE","0:00:00","Н/Д"
"svchost.exe","1548","Services","0","61 344 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:03","Н/Д"
"svchost.exe","1636","Services","0","11 992 КБ","Unknown","NT AUTHORITY\NETWORK SERVICE","0:00:00","Н/Д"
"svchost.exe","1684","Services","0","13 984 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"svchost.exe","1808","Services","0","20 232 КБ","Unknown","NT AUTHORITY\NETWORK SERVICE","0:00:00","Н/Д"
"svchost.exe","1940","Services","0","9 684 КБ","Unknown","NT AUTHORITY\LOCAL SERVICE","0:00:00","Н/Д"
"svchost.exe","2000","Services","0","27 448 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:03","Н/Д"
"svchost.exe","1612","Services","0","13 432 КБ","Unknown","NT AUTHORITY\LOCAL SERVICE","0:00:00","Н/Д"
"svchost.exe","1832","Services","0","10 268 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"Memory Compression","2308","Services","0","Н/Д","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"svchost.exe","2360","Services","0","9 160 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"svchost.exe","2392","Services","0","11 684 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"svchost.exe","2596","Services","0","10 296 КБ","Unknown","NT AUTHORITY\NETWORK SERVICE","0:00:00","Н/Д"
"svchost.exe","2624","Services","0","17 564 КБ","Unknown","NT AUTHORITY\LOCAL SERVICE","0:00:00","Н/Д"
"svchost.exe","2684","Services","0","12 128 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"svchost.exe","2760","Services","0","8 636 КБ","Unknown","NT AUTHORITY\LOCAL SERVICE","0:00:00","Н/Д"
"svchost.exe","2888","Services","0","24 204 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"spoolsv.exe","2980","Services","0","21 428 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"svchost.exe","3008","Services","0","21 780 КБ","Unknown","NT AUTHORITY\LOCAL SERVICE","0:00:00","Н/Д"
"svchost.exe","2528","Services","0","8 144 КБ","Unknown","NT AUTHORITY\LOCAL SERVICE","0:00:00","Н/Д"
"svchost.exe","3108","Services","0","14 740 КБ","Unknown","NT AUTHORITY\NETWORK SERVICE","0:00:00","Н/Д"
"svchost.exe","3116","Services","0","11 036 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"svchost.exe","3124","Services","0","9 876 КБ","Unknown","NT AUTHORITY\LOCAL SERVICE","0:00:00","Н/Д"
"OfficeClickToRun.exe","3212","Services","0","44 608 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"svchost.exe","3316","Services","0","9 716 КБ","Unknown","NT AUTHORITY\NETWORK SERVICE","0:00:00","Н/Д"
"svchost.exe","4012","Services","0","20 012 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:01","Н/Д"
"wlanext.exe","3468","Services","0","7 588 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"svchost.exe","4244","Services","0","11 376 КБ","Unknown","NT AUTHORITY\LOCAL SERVICE","0:00:00","Н/Д"
"sihost.exe","4412","","1","47 032 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"svchost.exe","4180","","1","22 800 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"svchost.exe","4548","","1","37 284 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"taskhostw.exe","4832","","1","17 604 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"svchost.exe","2776","Services","0","5 632 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"explorer.exe","5040","","1","246 828 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:07","Н/Д"
"ShellHost.exe","3356","","1","42 020 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"svchost.exe","5188","","1","21 460 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"SearchHost.exe","5500","","1","115 636 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:01","Н/Д"
"StartMenuExperienceHost.exe","5508","","1","153 824 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:03","Н/Д"
"CrossDeviceResume.exe","5592","","1","46 828 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"RuntimeBroker.exe","5692","","1","41 416 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"svchost.exe","5752","","1","27 860 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"SearchIndexer.exe","5836","Services","0","26 264 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"msedgewebview2.exe","6156","","1","110 604 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:01","Н/Д"
"msedgewebview2.exe","6428","","1","9 360 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"msedgewebview2.exe","6920","","1","52 708 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"msedgewebview2.exe","6976","","1","36 000 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"ctfmon.exe","7004","","1","36 424 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"msedgewebview2.exe","7136","","1","19 952 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"msedgewebview2.exe","6660","","1","88 908 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"WmiPrvSE.exe","7468","Services","0","17 900 КБ","Unknown","NT AUTHORITY\NETWORK SERVICE","0:00:00","Н/Д"
"taskhostw.exe","7244","","1","15 744 КБ","Unknown","DESKTOP-I8U3142\alice","0:00:00","Н/Д"
"svchost.exe","3948","Services","0","12 156 КБ","Unknown","NT AUTHORITY\LOCAL SERVICE","0:00:00","Н/Д"
"svchost.exe","7952","Services","0","17 248 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:05","Н/Д"
"svchost.exe","4852","Services","0","17 152 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"VSSVC.exe","7780","Services","0","10 580 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"svchost.exe","6024","Services","0","9 972 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"svchost.exe","668","Services","0","8 140 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"LogonUI.exe","4816","","1","95 160 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"csrss.exe","2832","Console","2","7 632 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"winlogon.exe","7764","Console","2","17 604 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"fontdrvhost.exe","5464","Console","2","5 408 КБ","Unknown","Font Driver Host\UMFD-2","0:00:00","Н/Д"
"dwm.exe","6916","Console","2","111 600 КБ","Running","Window Manager\DWM-2","0:00:00","DWM Notification Window"
"svchost.exe","4608","Services","0","9 844 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"sihost.exe","3024","Console","2","46 272 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","Н/Д"
"svchost.exe","7916","Console","2","22 372 КБ","Unknown","DESKTOP-I8U3142\jehy","0:00:00","Н/Д"
"svchost.exe","1712","Console","2","33 724 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","Windows Push Notifications Platform"
"taskhostw.exe","2696","Console","2","19 156 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","Task Host Window"
"explorer.exe","7896","Console","2","177 412 КБ","Running","DESKTOP-I8U3142\jehy","0:00:02","Н/Д"
"ShellHost.exe","6348","Console","2","40 952 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","Quick Settings"
"CrossDeviceResume.exe","3996","Console","2","46 668 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","CrossDeviceResumeWindow"
"svchost.exe","7236","Services","0","10 224 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"svchost.exe","4804","Console","2","21 428 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","OLEChannelWnd"
"StartMenuExperienceHost.exe","4536","Console","2","148 660 КБ","Running","DESKTOP-I8U3142\jehy","0:00:02","Пуск"
"SearchHost.exe","7248","Console","2","118 592 КБ","Running","DESKTOP-I8U3142\jehy","0:00:01","Поиск"
"RuntimeBroker.exe","1220","Console","2","42 736 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","Н/Д"
"svchost.exe","8132","Console","2","27 864 КБ","Unknown","DESKTOP-I8U3142\jehy","0:00:00","Н/Д"
"msedgewebview2.exe","3096","Console","2","108 912 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","Search"
"msedgewebview2.exe","5964","Console","2","9 428 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","Н/Д"
"msedgewebview2.exe","7928","Console","2","51 984 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","Temp Window"
"msedgewebview2.exe","6604","Console","2","35 412 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","OleMainThreadWndName"
"msedgewebview2.exe","7568","Console","2","19 268 КБ","Unknown","DESKTOP-I8U3142\jehy","0:00:00","Н/Д"
"msedgewebview2.exe","8288","Console","2","91 920 КБ","Unknown","DESKTOP-I8U3142\jehy","0:00:00","Н/Д"
"ctfmon.exe","8576","Console","2","26 916 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","Н/Д"
"audiodg.exe","8916","Services","0","15 836 КБ","Unknown","NT AUTHORITY\LOCAL SERVICE","0:00:00","Н/Д"
"cmd.exe","9004","Console","2","5 796 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","Администратор: C:\Windows\system32\cmd.exe - tasklist  /v /nh /fo csv"
"conhost.exe","9012","Console","2","19 204 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","Н/Д"
"chrome.exe","9084","Console","2","190 648 КБ","Running","DESKTOP-I8U3142\jehy","0:00:01","Telegram Web - Google Chrome"
"chrome.exe","9108","Console","2","9 008 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","Н/Д"
"chrome.exe","8524","Console","2","92 992 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","Н/Д"
"chrome.exe","8536","Console","2","42 556 КБ","Running","DESKTOP-I8U3142\jehy","0:00:00","OleMainThreadWndName"
"chrome.exe","4792","Console","2","20 836 КБ","Unknown","DESKTOP-I8U3142\jehy","0:00:00","Н/Д"
"chrome.exe","3916","Console","2","386 460 КБ","Unknown","DESKTOP-I8U3142\jehy","0:00:03","Н/Д"
"chrome.exe","9224","Console","2","18 176 КБ","Unknown","DESKTOP-I8U3142\jehy","0:00:00","Н/Д"
"chrome.exe","9264","Console","2","32 496 КБ","Unknown","DESKTOP-I8U3142\jehy","0:00:00","Н/Д"
"svchost.exe","9552","Services","0","11 832 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"svchost.exe","9588","Services","0","12 224 КБ","Unknown","NT AUTHORITY\LOCAL SERVICE","0:00:00","Н/Д"
"NgcIso.exe","9616","Services","0","3 892 КБ","Unknown","NT AUTHORITY\LOCAL SERVICE","0:00:00","Н/Д"
"taskhostw.exe","9796","Services","0","10 212 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"dllhost.exe","9836","Services","0","11 696 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"updater.exe","10172","Services","0","16 764 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"updater.exe","10192","Services","0","12 584 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"updater.exe","5936","Services","0","17 468 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"updater.exe","8776","Services","0","12 564 КБ","Unknown","NT AUTHORITY\СИСТЕМА","0:00:00","Н/Д"
"tasklist.exe","5376","Console","2","11 448 КБ","Unknown","DESKTOP-I8U3142\jehy","0:00:00","Н/Д"`;
    const res = task.parseReply(reply, true);
    expect(res).toMatchSnapshot()
  });
});
