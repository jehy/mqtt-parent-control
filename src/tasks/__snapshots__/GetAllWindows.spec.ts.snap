// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`GetAllWindows should be able to parse on WIN 1`] = `"Telegram Web - Google Chrome | Temp Window | Администратор: C:Windowssystem32cmd.exe - tasklist  /v /nh /fo csv"`;
